File: apps/admin/lib/core/contratos/despacho.dart
// El espejo en Dart de `packages/contratos/src/despacho.ts`. EP-07, ADR 019.
//
// Lo que el panel escribe al mover una Orden por el eje de entrega: por donde
// salio, por que no se entrego, por que se cancelo. Y las dos reglas que deciden
// que botones se ofrecen: que pasos escribe el panel (todos los de la tabla menos
// cancelar, que devuelve stock y lo hace el servidor) y cuando un pedido se puede
// despachar.
//
// `test/core/contratos/despacho_test.dart` compara contra
// `generated/contratos.json` (seccion `despacho`), donde las dos reglas viajan
// CALCULADAS por el TypeScript de hoy: el test no compara este archivo contra si
// mismo.

import 'estado_entrega.dart';
import 'estado_pago.dart';
import 'pedido.dart';

/// Por donde salio un pedido. Espejo de `CORREOS`: [clave] **es** el string que
/// queda en `despacho.correo`, y las reglas lo validan contra su propia copia.
enum Correo {
  andreani('andreani'),
  correoArgentino('correo_argentino'),
  oca('oca'),
  enMano('en_mano'),
  otro('otro');

  const Correo(this.clave);

  final String clave;

  static Correo? desde(Object? valor) {
    for (final c in values) {
      if (c.clave == valor) return c;
    }
    return null;
  }
}

/// Cuantos caracteres puede tener un numero de seguimiento. Espejo de
/// `LARGO_DEL_SEGUIMIENTO`. Es opcional: una entrega en mano no tiene.
const largoDelSeguimiento = 60;

/// Por que no se pudo entregar. Espejo de `MOTIVOS_DE_FALLA`. `sinMayor` es el
/// caso legal: `entregada` quiere decir que firmo un mayor de 18.
enum MotivoDeFalla {
  sinMayor('sin_mayor'),
  nadie('nadie'),
  direccion('direccion'),
  rechazo('rechazo'),
  otro('otro');

  const MotivoDeFalla(this.clave);

  final String clave;

  static MotivoDeFalla? desde(Object? valor) {
    for (final m in values) {
      if (m.clave == valor) return m;
    }
    return null;
  }
}

/// Por que se cancelo. Espejo de `MOTIVOS_DE_CANCELACION`.
enum MotivoDeCancelacion {
  malCargado('mal_cargado'),
  cliente('cliente'),
  faltaVino('falta_vino'),
  otro('otro');

  const MotivoDeCancelacion(this.clave);

  final String clave;

  static MotivoDeCancelacion? desde(Object? valor) {
    for (final m in values) {
      if (m.clave == valor) return m;
    }
    return null;
  }
}

/// Por que una linea de un pedido cancelado NO volvio al stock. Espejo de
/// `MOTIVOS_SIN_REPONER`.
enum MotivoSinReponer {
  noExiste('no_existe'),
  otraPresentacion('otra_presentacion'),
  compuesto('compuesto'),
  productoRoto('producto_roto');

  const MotivoSinReponer(this.clave);

  final String clave;

  static MotivoSinReponer? desde(Object? valor) {
    for (final m in values) {
      if (m.clave == valor) return m;
    }
    return null;
  }
}

/// Lo que el panel NUNCA escribe en `estadoEntrega`: cancelar devuelve el
/// stock, y eso lo hace la callable `cancelarOrden`. Espejo de
/// `SOLO_LO_ESCRIBE_EL_SERVIDOR`.
const soloLoEscribeElServidor = {EstadoEntrega.cancelada};

/// Si el panel escribe directo el paso de [antes] a [despues]. Espejo de
/// `laEscribeElPanel`: la tabla de ADR 002, menos lo del servidor, y sin
/// reescribir el mismo estado.
bool laEscribeElPanel(EstadoEntrega antes, EstadoEntrega despues) =>
    antes != despues &&
    !soloLoEscribeElServidor.contains(despues) &&
    transicionesEntrega[antes]!.contains(despues);

/// Un pedido de la vidriera impago no se despacha; uno de WhatsApp se despacha
/// sin mirar el pago, porque se cobra por fuera. Espejo de `sePuedeDespachar`.
bool sePuedeDespachar(Origen origen, EstadoPago estadoPago) =>
    origen == Origen.whatsapp || estadoPago == EstadoPago.pagada;

File: packages/contratos/src/despacho.ts
/**
 * Mover una Orden por el eje de entrega despues de que nacio: prepararla,
 * despacharla, registrar una entrega que fallo y cancelarla.  EP-07, ADR 019.
 *
 * Dos caminos, y la diferencia es de plata:
 *
 *  · Todo lo que NO toca el stock lo escribe el panel directo, y las reglas lo
 *    validan contra el documento de ahora (`resource.data`), sin una lectura
 *    extra: preparar, despachar, entregar y la entrega fallida.
 *  · Cancelar DEVUELVE el stock, y eso lo hace solo el servidor
 *    (`cancelarOrden`), en la misma transaccion que el cambio de estado.  Por eso
 *    `cancelada` esta en `SOLO_LO_ESCRIBE_EL_SERVIDOR`: si las reglas la dejaran
 *    escribir desde un cliente, un panel con un error cancelaria sin reponer, y
 *    el vino quedaria vendido para siempre.
 *
 * Este archivo es lo que los tres lados tienen que entender igual: la callable,
 * el panel (que lo espeja contra el JSON generado) y la suite de reglas (que lee
 * el MISMO JSON para saber que esperar).  Las reglas no pueden importar nada, asi
 * que tienen su propia copia de estas listas; la suite es lo que impide que esa
 * copia se desincronice.
 */

import { TRANSICIONES_ENTREGA, type EstadoEntrega, type EstadoPago, type Origen } from './orden.ts';
import type { Validacion } from './producto.ts';

// ----------------------------------------------------------------- despachar

/**
 * Por donde salio.  Se elige con un toque (poca burocracia, overview del panel).
 *
 * `en_mano` es *"lo llevamos nosotros"*: con el reparto propio apagado para la
 * vidriera (`REPARTIMOS_NOSOTROS = false`), un pedido de WhatsApp de la zona lo
 * puede llevar la familia.  `otro` existe para que ningun envio real quede sin
 * opcion y nadie elija uno falso; el nombre va en el seguimiento.
 * **Decision mia** (ADR 019 §4): el dueno la puede cambiar.
 */
export const CORREOS = ['andreani', 'correo_argentino', 'oca', 'en_mano', 'otro'] as const;

/**
 * El numero de seguimiento es OPCIONAL: una entrega en mano no tiene, y una
 * etiqueta se puede imprimir despues.  Cuando esta, de 1 a este largo: los
 * codigos de los correos argentinos rondan los 20 caracteres.
 */
export const LARGO_DEL_SEGUIMIENTO = 60;

/**
 * Un pedido de la VIDRIERA impago no se despacha: si Mercado Pago no aprobo, no
 * hay venta.  Uno de WHATSAPP se despacha sin mirar el pago, porque su cobro va
 * por fuera y el panel no lo sigue (EP-07, respuestas 5 y 9 del dueno).
 *
 * Las reglas lo repiten sobre `resource.data` con cero lecturas.
 */
export function sePuedeDespachar(origen: Origen, estadoPago: EstadoPago): boolean {
  return origen === 'whatsapp' || estadoPago === 'pagada';
}

// ------------------------------------------------------ la entrega que fallo

/**
 * Por que no se pudo entregar.  **`sin_mayor` es un caso operativo real, no un
 * adorno**: `entregada` significa que firmo un mayor de 18 (ARQUITECTURA §9.5),
 * asi que un paquete que llega a una casa con un menor solo no se entrega.
 */
export const MOTIVOS_DE_FALLA = ['sin_mayor', 'nadie', 'direccion', 'rechazo', 'otro'] as const;

// ------------------------------------------------------------------ cancelar

/** Lo que las reglas NUNCA dejan escribir desde un cliente en `estadoEntrega`. */
export const SOLO_LO_ESCRIBE_EL_SERVIDOR: readonly EstadoEntrega[] = ['cancelada'];

/**
 * Las transiciones que el panel escribe directo: las de la tabla, menos las que
 * son del servidor.  Reescribir el mismo estado NO es una transicion aca: el
 * panel lo hace al anotar, y eso no cambia nada del despacho.
 */
export function laEscribeElPanel(antes: EstadoEntrega, despues: EstadoEntrega): boolean {
  return (
    antes !== despues &&
    !SOLO_LO_ESCRIBE_EL_SERVIDOR.includes(despues) &&
    TRANSICIONES_ENTREGA[antes].includes(despues)
  );
}

export const MOTIVOS_DE_CANCELACION = ['mal_cargado', 'cliente', 'falta_vino', 'otro'] as const;
type MotivoDeCancelacion = (typeof MOTIVOS_DE_CANCELACION)[number];

/**
 * Por que una linea de una Orden cancelada NO volvio al stock.  Cancelar nunca
 * se traba por una linea: un pedido mal cargado que no se puede sacar cuesta
 * vino (ADR 018 §10).  Lo que no se puede devolver se dice, con su motivo, y lo
 * resuelve una persona.
 *
 *  · `no_existe`: el vino se borro.  No hay donde devolverlo.
 *  · `otra_presentacion`: el vino de hoy no viene en las botellas del snapshot
 *    -se borro y se volvio a crear con el mismo slug-.  Devolver 1 caja de 6
 *    como 1 botella suelta es el hallazgo 1 de ADR 008: 12 a la venta sobre 10
 *    fisicas.
 *  · `compuesto` / `producto_roto`: no tiene un stock propio sano donde sumar.
 */
export const MOTIVOS_SIN_REPONER = ['no_existe', 'otra_presentacion', 'compuesto', 'producto_roto'] as const;
export type MotivoSinReponer = (typeof MOTIVOS_SIN_REPONER)[number];

export interface LineaSinReponer {
  readonly productoId: string;
  readonly nombre: string;
  readonly cantidad: number;
  readonly motivo: MotivoSinReponer;
}

/** Lo que recibe `cancelarOrden`. */
export interface PedidoDeCancelacion {
  readonly ordenId: string;
  readonly motivo: MotivoDeCancelacion;
}

/** El mismo rango que `idPedido`: el `ordenId` ES el `idPedido` (ADR 018 §4). */
const ID_DE_ORDEN = /^[A-Za-z0-9_-]{16,64}$/;
const RESERVADO = /^__.*__$/;

function esObjeto(x: unknown): x is Record<string, unknown> {
  return typeof x === 'object' && x !== null && !Array.isArray(x);
}

/** RECHAZA lo que no cumple; no lo corrige.  Mismo trato que los otros parsers. */
export function parsearPedidoDeCancelacion(entrada: unknown): Validacion<PedidoDeCancelacion> {
  if (!esObjeto(entrada) || !Object.keys(entrada).every((k) => k === 'ordenId' || k === 'motivo')) {
    return { ok: false, motivo: 'el pedido no tiene la forma {ordenId, motivo}' };
  }
  if (typeof entrada.ordenId !== 'string' || !ID_DE_ORDEN.test(entrada.ordenId) || RESERVADO.test(entrada.ordenId)) {
    return { ok: false, motivo: 'ordenId invalido' };
  }
  if (typeof entrada.motivo !== 'string' || !(MOTIVOS_DE_CANCELACION as readonly string[]).includes(entrada.motivo)) {
    return { ok: false, motivo: `el motivo es uno de: ${MOTIVOS_DE_CANCELACION.join(', ')}` };
  }
  return { ok: true, valor: { ordenId: entrada.ordenId, motivo: entrada.motivo as MotivoDeCancelacion } };
}
